src/qililab/circuit_transpiler/circuit_transpiler.py

```diff
@@ -12,72 +12,161 @@
+        Virtual Z gates are also applied to correct phase errors from CZ gates.
+        The final RZ operator left to be applied as the last operator in the circuit can afterwards be removed since the last
+        operation is going to be a measurement, which is performed on the Z basis and is therefore invariant under rotations
+        around the Z axis.
+        This last step can also be seen from the fact that an RZ operator applied on a single qubit, with no operations carried
+        on afterwards induces a phase rotation. Since phase is an imaginary unitary component, its absolute value will be 1
+        independent on any (unitary) operations carried on it.
+
+        Mind that moving an operator to the left is equivalent to applying this operator last so
+        it is actually moved to the _right_ of Circuit.queue (last element of list).
+
+        For more information on virtual Z gates, see https://arxiv.org/abs/1612.00858
+
+        Args:
+            nqubits (int) : number of qubits in the circuit
+            ngates (list[gates.Gate]) : list of gates in the circuit
+
+        Returns:
+            list[gates.Gate] : list of re-ordered gates
+        """
+        supported_gates = ["rz", "drag", "cz", "wait", "measure"]
+        new_gates = []
+        shift = {qubit: 0 for qubit in range(nqubits)}
+        for gate in ngates:
+            if gate.name not in supported_gates:
+                raise NotImplementedError(f"{gate.name} not part of native supported gates {supported_gates}")
+            if isinstance(gate, gates.RZ):
+                shift[gate.qubits[0]] += gate.parameters[0]
+            # add CZ phase correction
+            elif isinstance(gate, gates.CZ):
+                gate_settings = self.platform.gates_settings.get_gate(name=gate.__class__.__name__, qubits=gate.qubits)
+                control_qubit, target_qubit = gate.qubits
+                corrections = next(
+                    (
+                        event.pulse.options
+                        for event in gate_settings
+                        if (
+                            event.pulse.options is not None
+                            and f"q{control_qubit}_phase_correction" in event.pulse.options
+                        )
+                    ),
+                    None,
+                )
+                if corrections is not None:
+                    shift[control_qubit] += corrections[f"q{control_qubit}_phase_correction"]
+                    shift[target_qubit] += corrections[f"q{target_qubit}_phase_correction"]
+                new_gates.append(gate)
+            else:
+                # if gate is drag pulse, shift parameters by accumulated Zs
+                if isinstance(gate, Drag):
+                    # create new drag pulse rather than modify parameters of the old one
+                    gate = Drag(gate.qubits[0], gate.parameters[0], gate.parameters[1] - shift[gate.qubits[0]])
```

Comment:
What is the difference between a new Drag or modifying the old one? Are we using the old one later one unchanged?

Reply:
This is mostly a safeguard. Every single qibo gate is a separate object with a single pointer to it. This is useful if you want to do things like setting parameters efficiently, but it also means that you can do the following:
```python
from qibo.gates import RX
from qibo.models import Circuit

c = Circuit(1)
c.add(RX(0,1))

c2 = Circuit(1)
rx = c.queue[0]
rx.parameters = (2,)
c2.add(rx)
print([gate.parameters for gate in c2.queue])
rx.parameters = (1,)
print([gate.parameters for gate in c2.queue])

out: [(2,)]
out: [(1,)]
```
So if I just add the original gate with the parameters changed, if someone then changes the parameters of the original circuit then the parameters of the transpiled circuit would also change.
There's also a `copy` method for qibo gates, but it's had bugs in the past so the best way I found of copying a gate with different parameters was to actually create a new gate with the desired parameters